src/backend/commands/matview.c

```diff
@@ -919,13 +927,16 @@ transientrel_init(QueryDesc *queryDesc)
 		tableSpace = matviewRel->rd_rel->reltablespace;
 		relpersistence = matviewRel->rd_rel->relpersistence;
 	}
+
+	ao_has_index = matviewRel->rd_rel->relhasindex && RelationIsAppendOptimized(matviewRel);
+
 	/*
 	 * Create the transient table that will receive the regenerated data. Lock
 	 * it against access by any other process until commit (by which time it
 	 * will be gone).
 	 */
 	OIDNewHeap = make_new_heap(matviewOid, tableSpace, relpersistence,
-							   ExclusiveLock, false, false);
+							   ExclusiveLock, ao_has_index, false);
```

Comment:
what about ExecuteTruncateGuts_IVM  ,line 3788 ?

Reply:
The current issue pertains to the use of append-optimized (AO) storage for materialized views, (IVM) do not support AO storage at this time.
```sql
ERROR: append-optimized table is not supported on incrementally maintainable materialized view.
```
Additionally, even if IVM were to support AO storage, why need to  handle truncate codes?